File: Stripe/NSURLComponents+Stripe.m
//
//  NSURLComponents+Stripe.m
//  Stripe
//
//  Created by Brian Dorfman on 1/26/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

#import "NSURLComponents+Stripe.h"

@implementation NSURLComponents (Stripe)

- (NSDictionary<NSString *, NSString *> *)stp_queryItemsDictionary {
    NSMutableDictionary *queryItems = [NSMutableDictionary new];

    for (NSURLQueryItem *queryItem in self.queryItems) {
        queryItems[queryItem.name] = queryItem.value;
    }

    return queryItems.copy;
}

- (void)setStp_queryItemsDictionary:(NSDictionary<NSString *, NSString *> *)stp_queryItemsDictionary {
    NSMutableArray *queryItems = [NSMutableArray new];
    [stp_queryItemsDictionary enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, NSString * _Nonnull obj, BOOL * __unused _Nonnull stop) {
        [queryItems addObject:[NSURLQueryItem queryItemWithName:key value:obj]];
    }];

    self.queryItems = queryItems.copy;
}

- (BOOL)stp_matchesURLComponents:(NSURLComponents *)rhsComponents {
    BOOL matches = ([[self.scheme lowercaseString] isEqualToString:[rhsComponents.scheme lowercaseString]]
                    && [[self.host lowercaseString] isEqualToString:[rhsComponents.host lowercaseString]]
                    && [self.path isEqualToString:rhsComponents.path]);

    if (matches) {
        NSDictionary<NSString *, NSString *> *rhsQueryItems = rhsComponents.stp_queryItemsDictionary;

        for (NSURLQueryItem *queryItem in self.queryItems) {
            if (![rhsQueryItems[queryItem.name] isEqualToString:queryItem.value]) {
                matches = NO;
                break;
            }
        }
    }

    return matches;
}

@end

void linkNSURLComponentsCategory(void){}

File: Stripe/NSURLComponents+Stripe.swift
//
//  NSURLComponents+Stripe.swift
//  Stripe
//
//  Created by Brian Dorfman on 1/26/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

import Foundation

extension NSURLComponents {
  /// Returns or sets self.queryItems as a dictionary where all the keys are the item
  /// names and the values are the values. When reading, if there are duplicate
  /// names, earlier ones are overwritten by later ones.
  @objc var stp_queryItemsDictionary: [String: String] {
    get {
      guard let queryItems = queryItems else {
        return [:]
      }
      var queryItemsDict = [String: String]()
      for queryItem in queryItems {
        queryItemsDict[queryItem.name] = queryItem.value
      }
      return queryItemsDict
    }
    set(stp_queryItemsDictionary) {
      var queryItems: [URLQueryItem] = []
      for (key, value) in stp_queryItemsDictionary {
        queryItems.append(URLQueryItem(name: key, value: value))
      }

      self.queryItems = queryItems
    }
  }

  /// Returns YES if the passed in url matches self in scheme, host, and path,
  /// AND all the query items in self are also in the passed
  /// in components (as determined by `stp_queryItemsDictionary`)
  /// This is used for URL routing style matching
  /// - Parameter rhsComponents: The components to match against
  /// - Returns: YES if there is a match, NO otherwise.
  func stp_matchesURLComponents(_ rhsComponents: NSURLComponents) -> Bool {
    var matches =
      (scheme?.lowercased() == rhsComponents.scheme?.lowercased())
      && (host?.lowercased() == rhsComponents.host?.lowercased()) && (path == rhsComponents.path)

    if matches {
      let rhsQueryItems = rhsComponents.stp_queryItemsDictionary

      for queryItem in queryItems ?? [] {
        if !(rhsQueryItems[queryItem.name] == queryItem.value) {
          matches = false
          break
        }
      }
    }

    return matches
  }
}
